src/metabase/notification/core.clj

```diff
@@ -56,26 +61,82 @@
              (.namingPattern "send-notification-thread-pool-%d"))))))
 
+(def ^:private default-retry-config
+  {:max-attempts            7
+   :initial-interval-millis 500
+   :multiplier              2.0
+   :randomization-factor    0.1
+   :max-interval-millis     30000})
+
+(defn- handler->channel-name
+  [{:keys [channel_type channel_id]}]
+  (if channel_id
+    (str (u/qualified-name channel_type) " " channel_id)
+    (u/qualified-name channel_type)))
+
+(defn- channel-send-retrying!
+  [handler message]
+  (try
+    (let [notification-id (:notification_id handler)
+          retry-config default-retry-config
+          retry-errors (volatile! [])
+          retry-report (fn []
+                         {:attempted_retries (count @retry-errors)
+                          ;; we want the last retry to be the most recent
+                          :retry_errors       (reverse @retry-errors)})
+          channel     (or (:channel handler)
+                          {:type (:channel_type handler)})
+          send!        (fn []
+                         (try
+                           (channel/send! channel message)
+                           (catch Exception e
+                             (vswap! retry-errors conj {:message   (u/strip-error e)
+                                                        :timestamp (t/offset-date-time)})
+                             (log/warnf e "[Notification %d] Failed to send to channel %s , retrying..." notification-id (handler->channel-name handler))
+                             (throw e))))]
+      (log/debugf "[Notification %d] Sending a message to channel %s" notification-id (handler->channel-name handler))
+      (task-history/with-task-history {:task            "channel-send"
+                                       :on-success-info (fn [update-map _result]
+                                                          (cond-> update-map
+                                                            (seq @retry-errors)
+                                                            (update :task_details merge (retry-report))))
+                                       :on-fail-info    (fn [update-map _result]
+                                                          (update update-map :task_details #(merge % (retry-report))))
+                                       :task_details    {:retry_config   retry-config
+                                                         :channel_id     (:id channel)
+                                                         :channel_type   (:type channel)
+                                                         :template_id    (:template_id handler)
+                                                         :notifcation_id notification-id
+                                                         :recipient_ids  (map :id (:recipients handler))}}
+        ((retry/decorate send! (retry/random-exponential-backoff-retry (str (random-uuid)) retry-config)))
```

Comment:
Kind of dislike this `decorate` pattern, I'd rather have something like `(let [retrier (retry/make <retry-args>)] (retrier send!))` — no confusing double parens. WDYT?

Reply:
yea haha, how about this [1ab084c](https://github.com/metabase/metabase/pull/48437/commits/1ab084c219e1c30d41601cd6f1a6a1aa4c9c8dd8)